compiler/sem/semfold.nim

```diff
@@ -730,32 +739,29 @@ proc getConstExpr(m: PSym, n: PNode; idgen: IdGenerator; g: ModuleGraph): PNode
         result.add a
     incl(result.flags, nfAllConst)
   of nkChckRangeF, nkChckRange64, nkChckRange:
-    var a = getConstExpr(m, n[0], idgen, g)
+    let a = getConstExpr(m, n[0], idgen, g)
     if a == nil: return
     if leValueConv(n[1], a) and leValueConv(a, n[2]):
       result = a              # a <= x and x <= b
       result.typ = n.typ
     else:
-      g.config.localReport(n.info, SemReport(
-        kind: rsemSemfoldInvalidConversion,
-        typeMismatch: @[typeMismatch(n[0].typ, n.typ)]))
-
+      result = g.config.newError(n,
+                                 PAstDiag(kind: adSemInvalidRangeConversion))
   of nkStringToCString, nkCStringToString:
-    var a = getConstExpr(m, n[0], idgen, g)
+    let a = getConstExpr(m, n[0], idgen, g)
     if a == nil: return
     result = a
     result.typ = n.typ
   of nkHiddenStdConv, nkHiddenSubConv, nkConv:
-    var a = getConstExpr(m, n[1], idgen, g)
+    let a = getConstExpr(m, n[1], idgen, g)
     if a == nil: return
     result = foldConv(n, a, idgen, g, check=true)
   of nkDerefExpr, nkHiddenDeref:
     let a = getConstExpr(m, n[0], idgen, g)
     if a != nil and a.kind == nkNilLit:
-      result = nil
-      #localReport(g.config, n.info, "nil dereference is not allowed")
+      result = nil   # xxx: why not return a nil lit instead?
```

Comment:
Hmm, I guess what's weird is that if we got an `nkNilLit` back, then this should in fact return an error (or we shouldn't have received an `nkNilLit`). I'm ok to leave it as is for now, as you've also indicated, there are going to be lots of fixes in this module and especially around `getConstExpr` callsites.

Hilariously, the error would be something like, `adSemConstDerefNilPointerException` 😆; I suspect in many cases this would be something we would recover from by aborting folding, but not necessarily always.

Reply:
Receiving an `nkNilLit` from folding is okay, because while a `nkNilLit` can't be folded itself (it's atomic), it's a node that the rest of folding can reason about and thus shouldn't block.

Having thought about the whole topic a bit more since my original comment, returning an error here would be fine, as long as it's later discarded when the folding is happening in the context of high-level optimization (i.e. invoked through `hlo -> commonOptimizations`).

Reply:
Left a note in the code saying that we should return `nkNilLit` and fix the code that's creating the incorrect deref.

Reply:
Hm, why is the deref incorrect? In following code:
```nim
const x = SomeRef(nil)[] # should fail with a nil dereference error, either during folding, or, like it does now, during VM execution

# `y` is replaced with `nkNilLit` during inlining:
if someCond:
  y
else:
  y[].other # should not lead to a compile-time error, except if executed at compile-time

```

there's are no incorrect derefs, and folding either of the `DerefExpr (NilLit)` expressions to just `NilLit` would result in semantically invalid code.

Reply:
Sorry about that, I was too hasty and what I wrote is not even close to what I was trying to convey. I'm pretty sure we're on the same page.

In case of `nkNiLit`, we should lift (not return) it and then produce an error because it's incorrect to dereference a `nil` -- example 1. Where `y` is speculatively replaced with `nkNilLit`, we should somehow inform `getConstExpr` to ignore certain `nkNilLit` nodes, and that should cover example 2 (inlining).

The parts where I meant we're generating incorrect derefs is focused on `nkHiddenDeref`. IIRC the analysis is not 100% and so I jumped to the conclusion we're misapplying them and a reason we end up in that code path.

Sorry for communicating all that poorly, thanks for the feedback and examples.